### PR TITLE
Initial implementation of the Maybe class.

### PR DESCRIPTION
**Pull request by jdantonio**
An implementation of the [Maybe](https://en.wikibooks.org/wiki/Haskell/Libraries/Maybe) pattern popularized by Haskell. This implementation is specifically intended to be used as the result of an operation that has failed with useful error information that must be communicated to the caller. If this isn't a good fit for the standard library I'd be happy to make this a shard instead.

Ruby overuses `nil`. Sometimes `nil` is a valid value, other times it represents the result of an operation that failed. This ambiguity can lead to difficult to understand and incorrect code. The [Null Object pattern](https://en.wikipedia.org/wiki/Null_Object_pattern) is one attempted solution, but it is overly complex. A new class or object must be defined for each class that may be null. When a Null Object is used the code generally just checks the type or identity, effectively just asking the object "did you succeed?" A `Maybe` is a generic solution to this problem. This `Maybe` is a specialization [option type](https://en.wikipedia.org/wiki/Option_type) intended to replace runtime exceptions.

This implementation of `Maybe` is based on a C++ class we use internally at work. It is an excellent fit for operations that would normally raise exceptions. Structured exception handling is considered an [anti-pattern](https://en.wikipedia.org/wiki/Anti-pattern) by many. Structured exception handling usurps the normal control flow, which can lead to unnecessarily complex code. This is the reason Erlang adopted the idiom of [returning a tuple](http://erlang.org/doc/man/erlang.html#match_spec_test-3) with the atoms `ok` or `error`, which became the basis for [Go's error handling](https://blog.golang.org/error-handling-and-go). Additionally, the caller of a method can never know _a priori_ if the method may raise an exception. In Crystal, neither can the compiler. Therefore the compiler cannot protect the programmer from unhandled runtime exceptions. Using a `Maybe` instead of raising exceptions solves all these problems.

When a method returns a `Maybe` the fact that an exception may occur is explicit. When the operation is complete there is no ambiguity regarding the result. The `Maybe` explicitly declares the success or failure of the operation. Moreover, the exact reason for the error is explicitly communicated, and the language's normal control flow can be used to handle the result.

This implementation is similar to my [Optional](https://github.com/crystal-lang/crystal/pull/3372) PR, but with several key differences. `Optional` is a generic solution to the null-value problem whereas `Maybe` is highly specialized for use with error handling. `Optional` objects are mutable whereas `Maybe` is not. `Maybe` solves a narrow problem, `Optional` has a wider range of uses.

```
require "json"

json_str = "[1, 2, 3]"
value = Maybe(JSON::Any).new do
  JSON.parse(json_str) # will succeed
end

value.succeeded? #=> true
value.failed?    #=> false
value.result     #=> [1, 2, 3]
value.failure    #=> nil

json_str = "not valid json"
value = Maybe(JSON::Any).new do
  JSON.parse(json_str) # will raise an exception
end

value.succeeded? #=> false
value.failed?    #=> true
value.result     #=> nil
value.failure    #=> #<JSON::ParseException:0x10b367d00 @message="unexpected char...
```

### CODE REVIEW

**Comment by fernandes** [comment]
hi @jdantonio , the build just broke because of coding format, run `crystal tool format` to fix it :)

never heard of Maybe pattern, interesting approach. can this be used as the foundation for promises implementation?

**Comment by jdantonio** [comment]
@fernandes Sorry about the build. I ran the tool and updated the PR.

WRT promises, yes, it can. Exceptions on threads don't bubble across threads, they crash the running thread. So when thread A sends a job to thread B, if there is an exception in thread B it will crash and thread A will never know what happened. So high-level concurrency libraries will rescue exceptions on the worker threads then pass the exception back to the caller threads using some type of structure. The calling thread can then decide what to do with it. Maybe is a prefect fit for this situation. I hadn't discovered Maybe when I began work on concurrent-ruby so we don't use them there, but I absolutely would if I started over today.

Another great use is for network operations since they are so error prone. Take a look at the API for the [HTTP::Request.from_io(io)](https://crystal-lang.org/api/0.19.3/HTTP/Request.html#from_io%28io%29-class-method) method. It has 3 possible return types:

* nil: EOF
* BadRequest: bad request
* HTTP::Request: successfully parsed

We could handle all of that with a simple `Maybe(HTTP::Request)` object. We would convey all the same information but have a single object. All of the if/then logic would be encapsulated in the `Maybe` so we'd likely have simpler calling code, too. Another benefit would be consistency. Right now we have to deal with potentially different return types for various methods across the standard library. If all such methods returned a `Maybe` there would be much greater consistency and developers new to the language would have an easier time learning the standard library.

Consistent, idiomatic code FTW! :-)

**Comment by lbguilherme** [comment]
First I would like to suggest an improvement to your implementation: Instead of storing two instance variables where they can never be both used at the same time, better use a single one with a union of types. This will be more compact memory-wise, specially if the result itself is already a union of types. Also removes the need to allow storing `nil` on there. Here is the (untested) code:

```cr
struct Maybe(T)
  struct Failure
    getter exception
    def initialize(@exception : Exception)
    end
  end
  
  def self.succeed(result : T)
    Maybe(T).new(result)
  end

  def self.fail(ex : Exception)
    Maybe(T).new(Failure.new(ex))
  end

  def self.fail(message : String)
    Maybe(T).new(Failure.new(Exception.new(message)))
  end

  protected def initialize(@result : T | Failure)
    @failure = nil
  end

  def succeeded?; !@result.is_a?(Failure); end
  def failed?; !succeeded?; end
  def message; failed? ? @result.as(Failure).exception.to_s : ""; end
  def result_or(value); succeeded? ? @result.as(T) : value; end
  def result_or; result_or(yield); end
end
```

---

About its inclusion on the standard library, I would oppose to it unless the standard library itself decide to use this idiom on several places where it is relevant. For most situations where the error message is not relevant, simply `nil` is used to replace it. For example: A `Channel` has a `receive?` method that returns `nil` when there is nothing to receive. Would that be changed to `Maybe(T)`? I think `nil` is simpler to handle and has compiler support (it is a falsy value).

**Comment by RX14** [comment]
This PR and #3372 seem to be PRs that oppose the current way things are done in both ruby and crystal. `nil` means "no value could be computed" by convention, it's even mentioned in [the docs](https://crystal-lang.org/api/0.19.3/Nil.html). The same is true with exceptions. The standard method of error handling is throwing exceptions and using `begin`/`rescue`/`end` blocks to handle them. Changing the standard practice to using `Maybe` and `Optional` would be huge departures from what I consider to be the core values of crystal: brevity by making the compiler work for you.

Handling a `Maybe` or `Option` type takes more code than the current methods because they take the concerns of the compiler (encoded in the type system) and encode them in a struct, which the user then has to handle. This seems like a huge backwards step to me.

When I write a method which raises an exception, I'm trusting the caller to handle the exception. If it doesn't the fiber crashes, which is sorta the caller's issue. `Maybe` also doesn't generalise to cases where you want to return a value but not an error. because you don't need a value from the `Maybe` class, you aren't encouraged to check for errors, so it's easy to silently ignore an exception, which is obviously a bad thing.

**Comment by asterite** [comment]
@jdantonio To add a bit more, for this to be useful we would probably have to remove exceptions from the language. In Go and Erlang returning a tuple is the only way to signal errors. In Rust it's the same, you return an `Error<T, R>` type (or something like that, I don't remember the exact name), and there are no exceptions.

Crystal embraces exceptions, which have good things and bad things:

1. (bad) Undinwing the stack is slow
2. (bad) Can't know which methods raise (but this is not true, the compiler can know this, and in fact the compiler right now knows which methods raise and which not, so we are only lacking the possible exception types)
3. (good) shorter code, no need to check and bubble error codes

For the `JSON.parse` case, a simple `JSON.parse(string) rescue nil` is enough. Of course `nil` could be the result of `JSON.parse`, because `"null"` is valid JSON, but nobody would send just that value in a JSON response because it conveys no information, so I think it's safe to do that. Another alternative is to have a `JSON.parse` that accepts a block that is only invoked when the string is valid.

`HTTP::Request.from_io(io)` is accidentally public, it's actually an internal implementation detail and we used this small union type because it works and it's simple, but if it were a public API we would maybe use a different union type.

In any case, I feel both `Optional` and `Maybe` have their use cases, but their inclusion might conflict a bit and bring a lot of confusion, so I'm hesitant about it.

**Comment by jdantonio** [comment]
@radarek "oppose the current way things are done in both ruby and crystal". Absolutely. As I mentioned in my other PR, Ruby isn't perfect. It's a great language but not everything in it is great. Nil handling and exceptions are two areas where I believe Ruby can do better. Crystal has an opportunity to critically evaluate Ruby and make changes where appropriate. So I thought I'd throw out some alternative ideas and see where the discussion went. So far it has been very informative and interesting. Hopefully others agree.

@asterite "Erlang returning a tuple is the only way to signal errors." This is incorrect. Erlang has [catch and throw](http://erlang.org/doc/reference_manual/expressions.html#catch) structured error handling. But they have fallen out of favor in place of returning tuples because the community feels that is the superior approach. So it's definitely possible to have both in a language simultaneously. One does not require the removal of the other. 

"but their inclusion might conflict a bit and bring a lot of confusion" This is definitely true, which is why I originally intended to make these a shard. To be perfectly honest, I'd be very happy if *every* language removed structured exception handling and replaced it with something like Erlang or Go, but I also know that I'm in the minority with this opinion. :-) I'm perfectly happy to create a shard with these classes if they don't fit in the standard library.

**Comment by rhysd** [comment]
IMO, this solution has the same defect as `std::optional` or `boost::optional` in C++.
From #3372,

> value.result

If the `Optional` class instance has `None` value, we still can access to `result` and it returns `nil`.  Why `Maybe` is awesome in Haskell is that there is a powerful pattern match and algebraic data types.

As discussed in above, we can use `T | Nil` for the purpose.  `T | Nil` is checked statically by compiler and unchecked `nil` causes an compilation error (great!).  If we want to have the exception information, we can use 'either' type such as `Result | Error`.  It's also checked by a compiler.

**Comment by ozra** [comment]
I know: I've mentioned this before, and it wasn't well received then, but: I still would like the approach where stdlib functions has both raising and Maybe-returning alternatives.
Imperative coding for me, is in part, for being _able_ to make really fuckin' fast stuff _when needed_.
- exceptions are great for situations where things rarely goes wrong, and we rather spend the cpu cycles only _then_, even though it's a murkier pattern (optimization paths only, imo).
- Maybes are great when the "operation failed" is more common _or_ a few extra cycles of checking value _each time_ is not a problem (so, most of code)

So Maybes does have the upper hand, and in my opinion exceptions are a "hackish" way of doing things that should be reserved for hot code paths with tight loops and such, where a lot of calls with return-value checking, for operations that are pretty much guaranteed to work (in the context), would be wasteful, and catching the entire hot-thing is better suited.

In short:
:+1: for Maybes as primary means of stdlib
:+1: for being able to use throwing versions when appropriate

Still just my 2c.

**Comment by cjgajard** [comment]
As it was already said, this can make Crystal tedious to write if it is included everywhere, but I'd like to see this, just like when (unsafe `[]` + safe `[]?`) was preferred over (safe `[]` + unsafe `[]!`), in a "safer" version of a method with `raise`, but returning a Maybe instead. It will be useful for dealing with "unsafe code you know is correct" with ease and to handle errors from untrusted sources easily too.

I imagine something like this:
```crystal
def positive(argument)
  raise ArgumentError if argument < 0
  argument
end

def positive!(argument)
end
```

The covention would be:
`foo` returns 
`foo!` returns a 
`foo?`

**Comment by trans** [comment]
This can be used as a required library -- nothing need actually be changing in Crystal itself, yes? If so then use it and promote it. I'm sure if enough people use it, it could migrate to Crystal proper. In the meantime you at least have what you want. I think it wouldn't be unreasonable to include in Crystal's standard library too, as opposed to core (are we using those Rubyisms?). That way it is easily available for all to try.

Having said that, it seems like your main concern is speed. Has anyone made an serious effort to figure out if their are ways to speed up the exception system?

Lastly. I've often wondered if there couldn't be a separate error channel that every method could use, kind of like stdout vs stderr. When doing a `return` one could add an optional error or even just a warning, and there would be some way to extract that from the returned value.

```ruby
    def foo
      return(nil){ BlahException.new("blah blah blah") }
    end
    x = foo
    x.error?  #=> true
    x.error    #=> BlahException : blah blah blah
```

So the error is just like little parasite that can stow a ride on any return value. I don't know, maybe it is a dumb idea. But it has been stick in the back of my mind for years.

**Comment by asterite** [comment]
I'm closing this issue. Like with #3417, a shard with Optional and Maybe can be provided.

**Comment by zatherz** [comment]
I think you linked the wrong PR.
